### ISSUE TITLE
Getting authentication error when running in Docker container

### ISSUE THREAD

**Issue by darrenparkinson** [koltyakov/sp-pnp-node#6]
I have an app in node which works perfectly, but when containerised using docker it gives the following error:
 
> "ID3242: The security token could not be authenticated or authorized."

Any thoughts on why that might be?

Many thanks,
D.

**Comment by koltyakov** [comment]
Hi Darren,

Can it be so that you copied `./config/private.json` file, which was used on another machine?

Password in the config is hashed with MachineId key and can't be decoded on the other machine. 
There is an option to disable encoding completely. 

P.S.
From the phone. Can check machineid situation in Docker in a few hours. 
I'm not very experinced with Docker. Should I ask you to share the dockerfile?

**Comment by darrenparkinson** [comment]
Hi, thanks for the reply.

Yes, I just discovered that [here](https://github.com/koltyakov/cpass/blob/master/src/utils/machineId.ts).  Although I also tried generating a new key on the container, but that also had some sort of issue. 

How can I test disabling the encoding easily? Just by setting config.encryptPassword to false?

**Comment by koltyakov** [comment]
Absolutely right, `config.encryptPassword = false` does it.

Some setup examples are:

```javascript
import * as pnp from 'sp-pnp-js';
import { PnpNode, IPnpNodeSettings } from 'sp-pnp-node';

let config = require('../config/private.json');

let pnpNodeSettings: IPnpNodeSettings = {
    siteUrl: config.siteUrl,
    authOptions: config,
    config: {
        encryptPassword: false
    }
};

pnp.setup({
    fetchClientFactory: () => {
        return new PnpNode(pnpNodeSettings);
    }
    // baseUrl: config.siteUrl
});

// PnP JS Core code...
```

or

```javascript
import { Web } from 'sp-pnp-js';
import { PnpNode, IPnpNodeSettings } from 'sp-pnp-node';

let pnpNodeSettings: IPnpNodeSettings = {
    config: {
        encryptPassword: false
    }
};

(new PnpNode(pnpNodeSettings)).initAmbient().then((settings) => {

    // PnP JS Core code...

}).catch(console.log);
```

The hashed password in existing private.json should be erased or replaced with raw password.

**Comment by koltyakov** [comment]
> Yes, I just discovered that [here](https://github.com/koltyakov/cpass/blob/master/src/utils/machineId.ts). Although I also tried generating a new key on the container, but that also had some sort of issue.

Does it failed there?
Are you running linux/windows container? 

Dockerfile, I assume, will be helpful for troubleshooting.

**Comment by darrenparkinson** [comment]
The container is linux.  Fairly simple docker file:

```Dockerfile
FROM node:8.1.3
WORKDIR /usr/src/app
COPY package.json /usr/src/app
RUN npm install
COPY . /usr/src/app
EXPOSE 3000
CMD ["node","index"]
```

**Comment by koltyakov** [comment]
Just tested [cpass](https://github.com/koltyakov/cpass) in Docker and it seems to have no issues:

```bash
PS D:\Projects\GitHub\cpass> docker run cpass
npm info it worked if it ends with ok
npm info using npm@5.0.3
npm info using node@v8.1.3
npm info lifecycle cpass@2.0.0~pretest: cpass@2.0.0
npm info lifecycle cpass@2.0.0~test: cpass@2.0.0

> cpass@2.0.0 test /usr/src/app
> mocha --opts test/mocha.opts || ECHO.

  Cpass tests
    Mode: Auto
      ✓ should encode a string
      ✓ should decode to the original
      ✓ should decode plain to itself
      ✓ should not decode modified hash
    Mode: Master Key
      ✓ should encode a string
      ✓ should decode to the original
      ✓ should decode plain to itself
      ✓ should not decode modified hash
    Special conditions
      ✓ should decode differently 1
      ✓ should decode differently 2
      ✓ should receive machineId
      ✓ should receive networkId
      ✓ should use machineId with priority
      ✓ should decode by a specific key


  14 passing (72ms)

npm info lifecycle cpass@2.0.0~posttest: cpass@2.0.0
npm info ok
```

What can I miss?

**Comment by koltyakov** [comment]
I've checked integration tests in Docker for sp-pnp-node too:

```bash
PS D:\Projects\GitHub\sp-pnp-node> docker run sp-pnp-node-test
npm info it worked if it ends with ok
npm info using npm@5.0.3
npm info using node@v8.1.3
npm info lifecycle sp-pnp-node@0.2.4~pretest: sp-pnp-node@0.2.4
npm info lifecycle sp-pnp-node@0.2.4~test: sp-pnp-node@0.2.4

> sp-pnp-node@0.2.4 test /usr/src/app
> ts-node ./test/init && mocha --opts test/mocha.opts || ECHO.

=== SharePoint Online Credentials ===

Gotcha /usr/src/app/config/integration/private.spo.json

=== On-Premise 2016 Credentials ===

Gotcha /usr/src/app/config/integration/private.2016.json

=== On-Premise 2013 Credentials ===

Gotcha /usr/src/app/config/integration/private.2013.json

  Run tests in SharePoint Online
    ...
    ✓ should create a new list (1896ms)
    ✓ should create list item (1567ms)
    ✓ should delete list item (1420ms)
    ✓ should fetch minimalmetadata (520ms)
    ✓ should fetch nometadata (508ms)
   ...
```

And it worked.
Private configs were setup with a plain passwords before docker image build.

**Comment by darrenparkinson** [comment]
Thankyou for testing that.  I got it to work by turning the encryption off, in a very similar way to your first example:

```js
let pnpNodeSettings: IPnpNodeSettings = {
    siteUrl: config.siteUrl,
    authOptions: config,
    config: {
        encryptPassword: false
    }
};
```
I have found it to be intermittent though with the account I'm using.  I'm having trouble elsewhere with the same account though, so I'll do some more testing, but it looks like the initial issue is fixed by removing the encryption which is tied to the machine id.

Thanks again for your prompt responses.
